File: java/src/main/java/io/github/nicklambourne/slackblocks/component/Paginator.java
package io.github.nicklambourne.slackblocks.component;

import io.github.nicklambourne.slackblocks.ErrorCategory;
import io.github.nicklambourne.slackblocks.ValidationException;
import io.github.nicklambourne.slackblocks.block.ActionsBlock;
import io.github.nicklambourne.slackblocks.block.Block;
import io.github.nicklambourne.slackblocks.block.ContextBlock;
import io.github.nicklambourne.slackblocks.element.ButtonElement;
import io.github.nicklambourne.slackblocks.object.MarkdownText;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import org.jspecify.annotations.Nullable;

/**
 * Renders one page of blocks plus previous and next buttons.
 *
 * <p>The buttons' action IDs are {@code <prefix>.previous} and {@code <prefix>.next}, and their
 * values hold the one-based page to show next, so an action handler can rebuild the message for
 * that page. Pass the result to the Slack SDK with {@code List.copyOf}, which widens it to {@code
 * List<LayoutBlock>}:
 *
 * <pre>{@code
 * List<Block> page = Paginator.builder("results").blocks(results).page(2).build();
 * client.chatPostMessage(ChatPostMessageRequest.builder()
 *     .channel("C0123456")
 *     .text("Search results")
 *     .blocks(List.copyOf(page))
 *     .build());
 * }</pre>
 */
public final class Paginator {
  private Paginator() {}

  /**
   * Starts a paginator with the prefix used for generated action identifiers.
   *
   * @param actionIdPrefix prefix for previous and next action IDs
   * @return a new builder
   */
  public static Builder builder(String actionIdPrefix) {
    return new Builder(actionIdPrefix);
  }

  /** Concrete fluent paginator builder. */
  public static final class Builder {
    private final String actionIdPrefix;
    private final List<Block> blocks = new ArrayList<>();
    private int page = 1;
    private int pageSize = 5;
    private String previousText = "Previous";
    private String nextText = "Next";
    private boolean showPageIndicator = true;
    private @Nullable String blockId;

    private Builder(String actionIdPrefix) {
      this.actionIdPrefix = Objects.requireNonNull(actionIdPrefix, "actionIdPrefix");
    }

    /**
     * Appends source blocks in display order.
     *
     * @param values source blocks
     * @return this builder
     */
    public Builder blocks(Block... values) {
      blocks.addAll(List.of(values));
      return this;
    }

    /**
     * Selects the one-based page to render.
     *
     * @param value one-based page number
     * @return this builder
     */
    public Builder page(int value) {
      page = value;
      return this;
    }

    /**
     * Sets the number of source blocks displayed per page.
     *
     * @param value positive page size
     * @return this builder
     */
    public Builder pageSize(int value) {
      pageSize = value;
      return this;
    }

    /**
     * Sets the previous-page button label.
     *
     * @param value button label
     * @return this builder
     */
    public Builder previousText(String value) {
      previousText = Objects.requireNonNull(value, "previousText");
      return this;
    }

    /**
     * Sets the next-page button label.
     *
     * @param value button label
     * @return this builder
     */
    public Builder nextText(String value) {
      nextText = Objects.requireNonNull(value, "nextText");
      return this;
    }

    /**
     * Controls whether a “Page n of m” context block is rendered.
     *
     * @param value whether to show the indicator
     * @return this builder
     */
    public Builder showPageIndicator(boolean value) {
      showPageIndicator = value;
      return this;
    }

    /**
     * Sets the identifier of the generated actions block. It is not used when every block fits on
     * one page, because no controls are rendered.
     *
     * @param value block identifier
     * @return this builder
     */
    public Builder blockId(String value) {
      blockId = Objects.requireNonNull(value, "blockId");
      return this;
    }

    /**
     * Builds the selected page followed, when there is more than one page, by a "Page n of m"
     * context block and the navigation buttons.
     *
     * @return immutable list of ordinary blocks
     * @throws ValidationException if there are no blocks, the prefix is empty, or the page or page
     *     size is out of range
     */
    public List<Block> build() {
      if (blocks.isEmpty()) {
        throw new ValidationException(
            ErrorCategory.MISSING_REQUIRED, "Paginator.blocks", "expected at least one block");
      }
      if (actionIdPrefix.isEmpty()) {
        throw new ValidationException(
            ErrorCategory.MISSING_REQUIRED,
            "Paginator.actionIdPrefix",
            "expected a non-empty prefix");
      }
      positive("Paginator.page", page);
      positive("Paginator.pageSize", pageSize);
      int pageCount = (blocks.size() + pageSize - 1) / pageSize;
      if (page > pageCount) {
        throw new ValidationException(
            ErrorCategory.OUT_OF_RANGE,
            "Paginator.page",
            "expected a value between 1 and " + pageCount);
      }
      int start = (page - 1) * pageSize;
      int end = Math.min(start + pageSize, blocks.size());
      List<Block> result = new ArrayList<>(blocks.subList(start, end));
      if (pageCount == 1) {
        return List.copyOf(result);
      }

      List<ButtonElement> controls = new ArrayList<>();
      if (page > 1) {
        controls.add(
            ButtonElement.builder(previousText, actionIdPrefix + ".previous")
                .value(Integer.toString(page - 1))
                .build());
      }
      if (page < pageCount) {
        controls.add(
            ButtonElement.builder(nextText, actionIdPrefix + ".next")
                .value(Integer.toString(page + 1))
                .build());
      }
      if (showPageIndicator) {
        result.add(
            ContextBlock.builder()
                .elements(MarkdownText.of("Page " + page + " of " + pageCount))
                .build());
      }
      ActionsBlock.Builder actions =
          ActionsBlock.builder().elements(controls.toArray(ButtonElement[]::new));
      String identifier = blockId;
      if (identifier != null) {
        actions.blockId(identifier);
      }
      result.add(actions.build());
      return List.copyOf(result);
    }

    private static void positive(String path, int value) {
      if (value < 1) {
        throw new ValidationException(
            ErrorCategory.OUT_OF_RANGE, path, "expected a positive integer");
      }
    }
  }
}

File: csharp/src/Slackblocks/Components/Paginator.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using Slackblocks.Blocks;
using Slackblocks.Elements;
using Slackblocks.Objects;

namespace Slackblocks.Components;

/// <summary>Selects one page of blocks and adds standard navigation controls.</summary>
/// <remarks>
/// The result is an ordinary list of blocks. When there is more than one page it ends with an
/// optional <c>Page N of M</c> context block and an actions block holding previous and next
/// buttons. Button action identifiers are <c>{actionIdPrefix}.previous</c> and
/// <c>{actionIdPrefix}.next</c>, and their values carry the one-based page number to show.
/// </remarks>
public static class Paginator
{
    /// <summary>Builds one page of <paramref name="blocks"/> with navigation.</summary>
    /// <param name="actionIdPrefix">Prefix for the navigation buttons' action identifiers.</param>
    /// <param name="blocks">Every block across all pages, in display order.</param>
    /// <param name="page">The one-based page to show.</param>
    /// <param name="pageSize">How many blocks each page shows.</param>
    /// <param name="previousText">Label of the previous-page button.</param>
    /// <param name="nextText">Label of the next-page button.</param>
    /// <param name="showPageIndicator">Whether to add a <c>Page N of M</c> context block.</param>
    /// <param name="blockId">Block identifier for the navigation actions block.</param>
    /// <returns>The blocks for the requested page, followed by the navigation blocks when needed.</returns>
    /// <exception cref="ValidationException">
    /// There are no blocks, the prefix is empty, or the page or page size is out of range.
    /// </exception>
    public static IReadOnlyList<IBlock> Create(
        string actionIdPrefix,
        IEnumerable<IBlock> blocks,
        int page = 1,
        int pageSize = 5,
        string previousText = "Previous",
        string nextText = "Next",
        bool showPageIndicator = true,
        string? blockId = null)
    {
        ArgumentNullException.ThrowIfNull(blocks);
        var all = blocks.ToList();
        if (all.Count == 0)
        {
            throw new ValidationException(ErrorCategory.MissingRequired, "Paginator.blocks", "expected at least one block");
        }

        if (string.IsNullOrEmpty(actionIdPrefix))
        {
            throw new ValidationException(ErrorCategory.MissingRequired, "Paginator.actionIdPrefix", "expected a non-empty prefix");
        }

        Positive("Paginator.page", page);
        Positive("Paginator.pageSize", pageSize);
        var pageCount = (all.Count + pageSize - 1) / pageSize;
        if (page > pageCount)
        {
            throw new ValidationException(
                ErrorCategory.OutOfRange,
                "Paginator.page",
                "expected a value between 1 and " + pageCount.ToString(CultureInfo.InvariantCulture));
        }

        var start = (page - 1) * pageSize;
        var result = all.Skip(start).Take(pageSize).ToList();
        if (pageCount == 1)
        {
            return result.AsReadOnly();
        }

        var controls = new List<ButtonElement>();
        if (page > 1)
        {
            controls.Add(new ButtonElement(
                previousText,
                actionIdPrefix + ".previous",
                value: (page - 1).ToString(CultureInfo.InvariantCulture)));
        }

        if (page < pageCount)
        {
            controls.Add(new ButtonElement(
                nextText,
                actionIdPrefix + ".next",
                value: (page + 1).ToString(CultureInfo.InvariantCulture)));
        }

        if (showPageIndicator)
        {
            result.Add(new ContextBlock([
                new MarkdownText(string.Create(CultureInfo.InvariantCulture, $"Page {page} of {pageCount}")),
            ]));
        }

        result.Add(new ActionsBlock(controls, blockId: blockId));
        return result.AsReadOnly();
    }

    private static void Positive(string path, int value)
    {
        if (value < 1)
        {
            throw new ValidationException(ErrorCategory.OutOfRange, path, "expected a positive integer");
        }
    }
}
